### PR TITLE
Port servicegraphprocessor to servicegraphconnector

### PR DESCRIPTION
**Pull request by djaglowski**
Similar to #18381. This is a copy/paste/update that results in a simpler codebase and config.

Still waiting for the framework to be merged into contrib before this can be fully validated, but it seemed like a pretty clean conversion.

### CODE REVIEW

**Comment by jpkrohling** [comment]
cc @kovrus and @mapno

**Review by mapno — COMMENTED** [review]
Thanks for the PR. TIL about connectors, and sound perfect for these types of components.

This PR will result in duplicated code bases, what are the following steps after this is merged? The processor gets marked as deprecated? I'd prefer to have both component at the same time for as short as possible.

**Comment by jpkrohling** [comment]
> This PR will result in duplicated code bases, what are the following steps after this is merged? The processor gets marked as deprecated? I'd prefer to have both component at the same time for as short as possible.

What I have talked to @kovrus (and perhaps @djaglowski, not sure), is to have both versions available for some time, allowing for a generous grace period for users to migrate over. The config migration should be easy, but not effortless: they are different component types even though the component's config is identical. We have not identified reasons for metrics to not be backward compatible, so that part should be covered.

**Comment by djaglowski** [comment]
I agree with @jpkrohling's suggested process, with one caveats.

As excited as I am about the connectors framework, I think we should exercise caution in the short term. The ability to use connectors will be included in this week's release, but will be disabled by default, behind an [`alpha` feature gate](https://github.com/open-telemetry/opentelemetry-collector/blob/main/featuregate/README.md#feature-lifecycle).

I think we can merge this whenever we are comfortable with the implementation but that we should not deprecate the processor and encourage migration until the connectors framework is enabled by default (`beta` stage of feature gate).

In other cases, I would suggest even more caution before recommending a migration. However, this processor is still `in development`, so requirements are minimal here.

**Comment by mapno** [comment]
What is was mainly worried about are both code bases diverging—eg. a contribution is being worked on in the processor ([PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17350#issuecomment-1422306256)), that as soon as its merged will already make this PR outdated.

From what I've checked `connector.Traces` and `consumer.Traces` are interchangeable. Would it be possible to export [`newProcessor`](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/servicegraphprocessor/processor.go#L83) and call it in the factory in this pkg? I think that's a low-effort compromise to make maintenance a bit simpler. I don't mean to block or slow the migration.

**Comment by djaglowski** [comment]
@mapno, seems reasonable. I'll try reuse what can be reused in the processor.

**Comment by djaglowski** [comment]
I reworked this so that the entire implementation remains in the processor. This seems ideal because we do not need to export anything additional.

The code ends up with some `if connector .. else ...` but it seems not too bad. I think it will be easy enough to condense back down to the connectors-only implementation when the time comes.

File: .chloggen/servicegraphconnector.yaml
Comment (kovrus) [inline .chloggen/servicegraphconnector.yaml:2]: new_component
```suggestion
change_type: new_component
```

File: connector/servicegraphconnector/README.md
Comment (kovrus) [inline connector/servicegraphconnector/README.md:9]: It mentions the `count` connector, we can probably remove it.

File: processor/servicegraphprocessor/factory_test.go
Comment (kovrus) [inline processor/servicegraphprocessor/factory_test.go:52]: nitpick: Maybe it is worth adding a separate test case or the test.

File: processor/servicegraphprocessor/config_test.go
Comment (kovrus) [inline processor/servicegraphprocessor/config_test.go:62]: Let's use `confmaptest` to load the component config, so then we can have only relevant to the component configuration in `testdata`.

File: connector/servicegraphconnector/README.md
Comment (djaglowski) [reply connector/servicegraphconnector/README.md:9]: Whoops, good catch.

File: processor/servicegraphprocessor/factory_test.go
Comment (djaglowski) [reply processor/servicegraphprocessor/factory_test.go:52]: I was on the fence about this one. I had to pick and choose what to do in parallel vs fully copy. I was on the fence about this one.

File: processor/servicegraphprocessor/config_test.go
Comment (djaglowski) [reply processor/servicegraphprocessor/config_test.go:62]: This test was copied from the one above. Do you think we should update both in this PR, or should this be a separate PR?

Comment (kovrus) [reply processor/servicegraphprocessor/config_test.go:62]: Imho, probably a separate PR would be better.

Comment (djaglowski) [reply processor/servicegraphprocessor/config_test.go:62]: Alright, I'll leave it as is then.

File: processor/servicegraphprocessor/factory_test.go
Comment (djaglowski) [reply processor/servicegraphprocessor/factory_test.go]: Done

File: connector/servicegraphconnector/README.md
Comment (djaglowski) [reply connector/servicegraphconnector/README.md]: Removed

File: processor/servicegraphprocessor/processor_test.go
Comment (kovrus) [inline processor/servicegraphprocessor/processor_test.go:98]: Do we need to call shutdown after start to be safe?

**Review by kovrus — COMMENTED** [review]
lgtm! 

minor, some of the logs contain the `servicegraphprocessor` name which might be confusing for connector users.

**Comment by djaglowski** [comment]
I updated the log messages to account for the component type

**Comment by djaglowski** [comment]
I think this should be good to go, unless there is more feedback. @jpkrohling, do you want to take a look?

**Review by kovrus — APPROVED** [review]

File: .chloggen/servicegraphconnector.yaml
Comment (djaglowski) [reply .chloggen/servicegraphconnector.yaml]: I realized I forgot this. Fixed now.

**Pull request merged**
